File: runme/mcgen/local/datan/datsrc/lsqnon.c
#include <stdlib.h>
#include <limits.h>
#include "datsrc.h"

/* Common Block Declarations */

/*
struct {
    double c[1000], x2[1000], x1red[1000];
} dasv02_;

#define dasv02_1 dasv02_
*/

void LIBFUNC lsqnon_(double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *, double const P_T *), double const P_T *t,
             double const P_T *y, double const P_T *deltay,
             integer const P_T *n, integer const P_T *nr, integer P_T *nred,
             integer P_T *list, double P_T *x, double P_T *cx, double P_T *r,
             double P_T *a, double P_T *scrat, integer P_T *nstep)
{
    /* Table of constant values */

    integer c__1 = 1;
    double c_d0 = 0.;
    double c_dp5 = .5;

    /* System generated locals */
    integer a_dim1, a_offset, scrat_dim1, scrat_offset, cx_dim1, cx_offset,
            i__1, i__2, i__3;
    double d__1;

    /* Local variables */
    integer i, k, istep;
    double *c, *x2, *x1red, r1;
    logical ok, covmat;
    unsigned long amem;

    if ((amem = sizeof(double) * *n) > MAXALLOC) {
        MEMERROR("lsqnon_: c (too large)");
        goto f_0;
    }
    c = (double *)MEMALLOC(amem);
    if (!c) {
        MEMERROR("lsqnon_: c");
        goto f_0;
    }
    if ((amem = sizeof(double) * (*nr + *nred)) > MAXALLOC) {
        MEMERROR("lsqnon_: x2 (too large)");
        goto f_1;
    }
    x2 = (double *)MEMALLOC(amem);
    if (!x2) {
        MEMERROR("lsqnon_: x2");
        goto f_1;
    }
    x1red = x2 + *nr;

    /* Parameter adjustments */
    --deltay;
    --y;
    --t;
    --x;
    --list;
    scrat_dim1 = *nred;
    scrat_offset = scrat_dim1 + 1;
    scrat -= scrat_offset;
    a_dim1 = *n;
    a_offset = a_dim1 + 1;
    a -= a_offset;
    cx_dim1 = *nred;
    cx_offset = cx_dim1 + 1;
    cx -= cx_offset;

    /* Function Body */
    covmat = TRUE_;
    if (*nstep < 0) {
        covmat = FALSE_;
        *nstep = abs(*nstep);
    }
    if (*nstep < 1) {
        *nstep = 100;
    }
/* Initial value of minimum function */
    *r = 0.;
    i__1 = *n;
    for (i = 1; i <= i__1; ++i) {
/* Computing 2nd power */
        d__1 = (y[i] - (*userfn)(&x[1], nr, &t[i])) / deltay[i];
        *r += d__1 * d__1;
    }
/* For NRED=0 : set NSTEP=1 and return */
    if (*nred <= 0) {
        *nstep = 1;
        goto L130;
    }
/* For NR=NRED :  set LIST */
    if (*nr == *nred) {
        i__1 = *nr;
        for (i = 1; i <= i__1; ++i) {
            list[i] = 1;
        }
    }
/* start iteration */
    i__1 = *nstep;
    for (istep = 1; istep <= i__1; ++istep) {
/* Numerical Derivatives */
        auxdri_(userfn, &x[1], &t[1], n, nr, nred, &list[1], &a[a_offset], &ok);
        if (! ok) {
            *nstep = -3;
            goto L130;
        }
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
            i__3 = *nred;
            for (k = 1; k <= i__3; ++k) {
                a[i + k * a_dim1] /= deltay[i];
            }
        }
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
            c[i - 1] = (y[i] - (*userfn)(&x[1], nr, &t[i])) / deltay[i];
        }
        mtxsvd_(&a[a_offset], c, x1red, &r1, n, nred, &c__1,&c_d0, &ok);
        if (! ok) {
            *nstep = -1;
            goto L130;
        }
        mtxzrv_(x2, nr);
        mtxpsv_(x2, x1red, nr, nred, &list[1]);
        mtxadv_(&x[1], x2, x2, nr);
        r1 = 0.;
        i__2 = *n;
        for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
            d__1 = (y[i] - (*userfn)(x2, nr, &t[i])) / deltay[i];
            r1 += d__1 * d__1;
        }
        for (k = 1; k <= 30; ++k) {
            if (r1 > *r * 1.00000001 + 1e-15) {
/* minimum function has increased: reduce step by factor of two */
                mtxmsv_(x1red, x1red, &c_dp5, nred);
                mtxzrv_(x2, nr);
                mtxpsv_(x2, x1red, nr, nred, &list[1]);
                mtxadv_(&x[1], x2, x2, nr);
                r1 = 0.;
                i__2 = *n;
                for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
                    d__1 = (y[i] - (*userfn)(x2, nr, &t[i])) / deltay[i];
                    r1 += d__1 * d__1;
                }
            } else {
                goto L90;
            }
        }
        *nstep = -4;
        goto L130;
L90:
        mtxzrv_(x2, nr);
        mtxpsv_(x2, x1red, nr, nred, &list[1]);
        mtxadv_(&x[1], x2, &x[1], nr);
/* test for break-off criterion */
        if ((d__1 = *r - r1, abs(d__1)) < abs(*r) * 1e-8 + 1e-15) {
            *r = r1;
/* Compute covariance matrix CX of unknowns */
            if (covmat) {
                auxdri_(userfn, &x[1], &t[1], n, nr, nred, &list[1],
                        &a[a_offset], &ok);
                if (! ok) {
                    *nstep = -3;
                    goto L130;
                }
                i__2 = *n;
                for (i = 1; i <= i__2; ++i) {
                    i__3 = *nred;
                    for (k = 1; k <= i__3; ++k) {
                        a[i + k * a_dim1] /= deltay[i];
                    }
                }
                mtxmat_(&a[a_offset], &a[a_offset], &cx[cx_offset], nred, n,
                        nred);
                mtxchi_(&cx[cx_offset], &scrat[scrat_offset], nred);
                *nstep = istep;
            }
            goto L130;
        } else {
            *r = r1;
        }
    }
    *nstep = -2;
L130:
    MEMFREE((void *)x2);
f_1:
    MEMFREE((void *)c);
f_0:
    return;
} /* lsqnon_ */

File: runme/mcgen/local/datan/datsrc/lsqnon.f
      SUBROUTINE LSQNON(USERFN,T,Y,DELTAY,N,NR,NRED,LIST,X,CX,R,A,
     +SCRAT,NSTEP)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(NMAX=1000)
      DIMENSION A(N,NRED),T(N),Y(N),DELTAY(N),X(NR)
      DIMENSION SCRAT(NRED,NRED),CX(NRED,NRED),C(NMAX)
      DIMENSION X2(NMAX),X1RED(NMAX)
      DIMENSION LIST(NR)
      COMMON /DASV02/ C,X2,X1RED
      PARAMETER(MAXSTP=100,EPSILN=1.D-8,TT=1.D-15,FACT=0.5D0)
      PARAMETER(ZERO=0.0D0)
      LOGICAL OK,COVMAT
      EXTERNAL USERFN
      COVMAT=.TRUE.
      IF(NSTEP.LT.0) THEN
        COVMAT=.FALSE.
        NSTEP=ABS(NSTEP)
      END IF
      IF(NSTEP.LT.1) NSTEP=MAXSTP
C Initial value of minimum function
      R=ZERO
      DO 10 I=1,N
        R=R+((Y(I)-USERFN(X,NR,T(I)))/DELTAY(I))**2
   10 CONTINUE
C For NRED=0 : set NSTEP=1 and return
      IF(NRED.LE.0) THEN
        NSTEP=1
        GO TO 130
      END IF
C For NR=NRED :  set LIST
      IF(NR.EQ.NRED) THEN
        DO 20 I=1,NR
          LIST(I)=1
   20   CONTINUE
      END IF
C start iteration
      DO 120 ISTEP=1,NSTEP
C Numerical Derivatives
        CALL AUXDRI(USERFN,X,T,N,NR,NRED,LIST,A,OK)
        IF(.NOT.OK) THEN
          NSTEP=-3
          GO TO 130
        END IF
        DO 40 I=1,N
          DO 30 K=1,NRED
            A(I,K)=A(I,K)/DELTAY(I)
   30     CONTINUE
   40   CONTINUE
        DO 50 I=1,N
          C(I)=(Y(I)-USERFN(X,NR,T(I)))/DELTAY(I)
   50   CONTINUE
        CALL MTXSVD(A,C,X1RED,R1,N,NRED,1,DBLE(0.),OK)
        IF(.NOT.OK) THEN
          NSTEP=-1
          GO TO 130
        END IF
        CALL MTXZRV(X2,NR)
        CALL MTXPSV(X2,X1RED,NR,NRED,LIST)
        CALL MTXADV(X,X2,X2,NR)
        R1=ZERO
        DO 60 I=1,N
          R1=R1+((Y(I)-USERFN(X2,NR,T(I)))/DELTAY(I))**2
   60   CONTINUE
        DO 80 K=1,30
          IF(R1.GT.R*(DBLE(1.)+EPSILN)+TT) THEN
C minimum function has increased: reduce step by factor of two
            CALL MTXMSV(X1RED,X1RED,FACT,NRED)
            CALL MTXZRV(X2,NR)
            CALL MTXPSV(X2,X1RED,NR,NRED,LIST)
            CALL MTXADV(X,X2,X2,NR)
            R1=ZERO
            DO 70 I=1,N
              R1=R1+((Y(I)-USERFN(X2,NR,T(I)))/DELTAY(I))**2
   70       CONTINUE
          ELSE
            GO TO 90
          END IF
   80   CONTINUE
        NSTEP=-4
        GO TO 130
   90   CONTINUE
        CALL MTXZRV(X2,NR)
        CALL MTXPSV(X2,X1RED,NR,NRED,LIST)
        CALL MTXADV(X,X2,X,NR)
C test for break-off criterion
        IF(ABS(R-R1) .LT. EPSILN*ABS(R)+TT) THEN
          R=R1
C Compute covariance matrix CX of unknowns
          IF(COVMAT) THEN
            CALL AUXDRI(USERFN,X,T,N,NR,NRED,LIST,A,OK)
            IF(.NOT.OK) THEN
              NSTEP=-3
              GO TO 130
            END IF
            DO 110 I=1,N
              DO 100 K=1,NRED
                A(I,K)=A(I,K)/DELTAY(I)
  100         CONTINUE
  110       CONTINUE
            CALL MTXMAT(A,A,CX,NRED,N,NRED)
            CALL MTXCHI(CX,SCRAT,NRED)
            NSTEP=ISTEP
          END IF
          GO TO 130
        ELSE
          R=R1
        END IF
  120 CONTINUE
      NSTEP=-2
  130 RETURN
      END
